src/Evolution/Systems/Cce/BoundaryData.cpp

```diff
@@ -0,0 +1,559 @@
+    const tnsr::iJ<DataVector, 3>& d_cartesian_shift,
+    const tnsr::I<DataVector, 3>& dt_cartesian_shift,
+    const Scalar<DataVector>& cartesian_lapse,
+    const tnsr::i<DataVector, 3>& d_cartesian_lapse,
+    const Scalar<DataVector>& dt_cartesian_lapse) noexcept {
+  dt_psi->destructive_resize_components(get(dt_cartesian_lapse).size());
+
+  GeneralizedHarmonic::phi(
+      phi, cartesian_lapse, d_cartesian_lapse, cartesian_shift,
+      d_cartesian_shift, cartesian_spatial_metric, d_cartesian_spatial_metric);
+
+  get<0, 0>(*dt_psi) = -2.0 * get(cartesian_lapse) * get(dt_cartesian_lapse);
+  for (size_t i = 0; i < 3; ++i) {
+    for (size_t j = 0; j < 3; ++j) {
+      get<0, 0>(*dt_psi) +=
+          dt_cartesian_spatial_metric.get(i, j) * cartesian_shift.get(i) *
+              cartesian_shift.get(j) +
+          2.0 * cartesian_spatial_metric.get(i, j) * cartesian_shift.get(i) *
+              dt_cartesian_shift.get(j);
+    }
+  }
+
+  for (size_t i = 0; i < 3; ++i) {
+    dt_psi->get(0, i + 1) = 0.0;
+    for (size_t j = 0; j < 3; ++j) {
+      dt_psi->get(0, i + 1) +=
+          dt_cartesian_spatial_metric.get(i, j) * cartesian_shift.get(j) +
+          cartesian_spatial_metric.get(i, j) * dt_cartesian_shift.get(j);
+    }
+    dt_psi->get(i + 1, 0) = dt_psi->get(0, i + 1);
+  }
+
+  for (size_t i = 0; i < 3; ++i) {
+    for (size_t j = i; j < 3; ++j) {
+      dt_psi->get(i + 1, j + 1) = dt_cartesian_spatial_metric.get(i, j);
+    }
+  }
+
+  gr::spacetime_metric(psi, cartesian_lapse, cartesian_shift,
+                       cartesian_spatial_metric);
+}
+
+void null_metric_and_derivative(
+    const gsl::not_null<tnsr::aa<DataVector, 3>*> du_null_metric,
+    const gsl::not_null<tnsr::aa<DataVector, 3>*> null_metric,
+    const tnsr::iJ<DataVector, 3>& cartesian_to_spherical_jacobian,
+    const tnsr::aa<DataVector, 3>& dt_psi,
+    const tnsr::aa<DataVector, 3>& psi) noexcept {
+  size_t size = get<0, 0>(psi).size();
+  null_metric->destructive_resize_components(size);
+  du_null_metric->destructive_resize_components(size);
+
+  get<0, 0>(*null_metric) = get<0, 0>(psi);
+  get<0, 0>(*du_null_metric) = get<0, 0>(dt_psi);
+
+  get<0, 1>(*null_metric) = -1.0;
+  get<0, 1>(*du_null_metric) = 0.0;
+
+  for (size_t i = 0; i < 3; ++i) {
+    null_metric->get(1, i + 1) = 0.0;
+    du_null_metric->get(1, i + 1) = 0.0;
+  }
+
+  for (size_t A = 0; A < 2; ++A) {
+    null_metric->get(0, A + 2) =
+        cartesian_to_spherical_jacobian.get(A + 1, 0) * psi.get(0, 1);
+    du_null_metric->get(0, A + 2) =
+        cartesian_to_spherical_jacobian.get(A + 1, 0) * dt_psi.get(0, 1);
+    for (size_t i = 1; i < 3; ++i) {
+      null_metric->get(0, A + 2) +=
+          cartesian_to_spherical_jacobian.get(A + 1, i) * psi.get(0, i + 1);
+      du_null_metric->get(0, A + 2) +=
+          cartesian_to_spherical_jacobian.get(A + 1, i) * dt_psi.get(0, i + 1);
+    }
+  }
+
+  for (size_t A = 0; A < 2; ++A) {
```

Comment:
Yikes... I think what you're pointing to right here would probably generalize reasonably gracefully, but I'd be concerned that incorporating that generality in the code might make someone think we have any idea how to do CCE in other dimensions, and I'm pretty certain that basic things like how we treat the angular basis vectors would be super wrong in anything apart from spatial dimension 3. Feel free to assign me an issue to think more carefully about what subset of these utilities would be 'graceful' and which would break horribly, and to generalize the graceful ones.